envoy-control-core/src/main/java/pl/allegro/tech/servicemesh/envoycontrol/SimpleCache.java

```diff
@@ -0,0 +1,351 @@
+     */
+    @Override
+    public Snapshot getSnapshot(T group) {
+        readLock.lock();
+
+        try {
+            return snapshots.get(group);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Collection<T> groups() {
+        return ImmutableSet.copyOf(statuses.keySet());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void setSnapshot(T group, Snapshot snapshot) {
+        // we take a writeLock to prevent watches from being created while we update the snapshot
+        CacheStatusInfo<T> status;
+        writeLock.lock();
+        try {
+            // Update the existing snapshot entry.
+            snapshots.put(group, snapshot);
+            status = statuses.get(group);
+        } finally {
+            writeLock.unlock();
+        }
+
+        if (status == null) {
+            return;
+        }
+
+        for (String typeUrl : Resources.TYPE_URLS) {
+            status.watchesRemoveIf((id, watch) -> {
+                if (!watch.request().getTypeUrl().equals(typeUrl)) {
+                    return false;
+                }
+                String version = snapshot.version(watch.request().getTypeUrl(), watch.request().getResourceNamesList());
+
+                if (!watch.request().getVersionInfo().equals(version)) {
+                    if (LOGGER.isDebugEnabled()) {
+                        LOGGER.debug("responding to open watch {}[{}] with new version {}",
+                                id,
+                                String.join(", ", watch.request().getResourceNamesList()),
+                                version);
+                    }
+
+                    respond(watch, snapshot, group);
+
+                    // Discard the watch. A new watch will be created for future snapshots once envoy ACKs the response.
+                    return true;
+                }
+
+                // Do not discard the watch. The request version is the same as the snapshot version, so we wait to respond.
+                return false;
+            });
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public StatusInfo statusInfo(T group) {
+        readLock.lock();
+
+        try {
+            return statuses.get(group);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    private Response createResponse(DiscoveryRequest request, Map<String, ? extends Message> resources, String version) {
+        Collection<? extends Message> filtered = request.getResourceNamesList().isEmpty()
+                ? resources.values()
+                : request.getResourceNamesList().stream()
+                .map(resources::get)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        return Response.create(request, filtered, version);
+    }
+
+    private boolean respond(Watch watch, Snapshot snapshot, T group) {
```

Comment:
Could you please split this method? There's a lot going on here making in hard to read and understand.

Reply:
I want to keep it as close as possible to class existing in `java-control-plane`.